### ISSUE TITLE
`reroute` enhancements

### ISSUE THREAD

**Issue by benmccann** [sveltejs/kit#11595]
### Describe the problem

Content will be available at two URLs when using `reroute`. E.g. https://kit.svelte.dev/docs and https://kit.svelte.dev/en/docs. This isn't bad for user experience and SEO. You can address it by redirecting one URL like https://kit.svelte.dev/en/docs to https://kit.svelte.dev/docs but it's kind of clunky if you need to updated two places everytime you want to use `reroute`. Though one benefit of doing it in `handle` is that you can choose what type of redirect you want

People have also mentioned wanting to `reroute` using cookies, which the current API doesn't allow. E.g. you might put someone into an A/B test bucket and want to show them the same version of the page everytime. If it's just content it's easier to A/B test in the template with an `if` statement, but if you need to load different data then something like rerouting might be beneficial

### Describe the proposed solution

Perhaps think about whether/if we'd like to update the API at all as part of 3.0. We could also think about whether we want these files to live in a `src/kit` directory or change the `.client`/`.server` portion of the filename, etc.

### Importance

nice to have

**Comment by paoloricciuti** [comment]
Isn't the fact that the content is the same in two routes basically the main point of reroute? I want to have the same content in `/en/something` and in `/it/qualcosa` but i just want `/[lang]/something` in my codebase.

**Comment by f-elix** [comment]
From what I've seen it doesn't have to be that way. You could just have a `[lang]/docs` route, where the lang param is not optional, so that `/docs` returns a 404.

**Comment by benmccann** [comment]
> Isn't the fact that the content is the same in two routes basically the main point of reroute? I want to have the same content in /en/something and in /it/qualcosa but i just want /[lang]/something in my codebase.

By same content, I mean same exact content - i.e. same language. In this example there's an optional lang parameter and I don't want people to be able to access the English version at two different URLs, but just a single URL.

> From what I've seen it doesn't have to be that way. You could just have a [lang]/docs route, where the lang param is not optional, so that /docs returns a 404.

Sure, there are many ways to setup i18n, but I don't want to turn this into a thread about that. This issue is about how the `reroute` API should look.

**Comment by paoloricciuti** [comment]
> > Isn't the fact that the content is the same in two routes basically the main point of reroute? I want to have the same content in /en/something and in /it/qualcosa but i just want /[lang]/something in my codebase.
> 
> By same content, I mean same exact content - i.e. same language. In this example there's an optional lang parameter and I don't want people to be able to access the English version at two different URLs, but just a single URL.
> 
> > From what I've seen it doesn't have to be that way. You could just have a [lang]/docs route, where the lang param is not optional, so that /docs returns a 404.
> 
> Sure, there are many ways to setup i18n, but I don't want to turn this into a thread about that. This issue is about how the `reroute` API should look.

Do you think that should be the job for `reroute`? I think that should be a concern of the developer. So once I setup `reroute` I should also make sure that the two routes are different in content (or maybe not because I want to have two routes with the same content for some reason)

**Comment by dominikg** [comment]
+1 to reroute with headers. It's not just A/B testing but maybe language stored in a cookie or wanting to use the Accept-Lanugage header. 

It can work by providing the headers only on serverside in reroute. Client can get the language from `<html lang="XX">`  which should have been produced by SSR or prerendering.


```js
import {browser} from '$app/environment';
export const reroute ({url,headers}) {
  const lang = (browser ? document.documentElement.lang : getLang({url,headers})) ?? defaultLang;
  return getRoutePath(url,lang); // localization mapping in here
}
```

**Comment by dominikg** [comment]
Additionally, reroute currently only provides a one-way mapping of an incoming url to a route, but usually you need the opposite direction too to be able to produce matching hrefs during ssr.

inlang/paraglide currently has a sveltekit adapter that uses a hardcoded [lang] in first path segment scheme, but this does not work for everyone. 

To avoid lots of different userland implementations doing all the legwork it would be great if kit had an api that was similar to atob btoa where you can have both ways set.

**Comment by dominikg** [comment]
If reroute is used for localized routes, should it be easily possible to disallow the target route? 
eg if you have a localized about page in `src/routes/about/+page.svelte`  and reroute `/about` to `/ueber-uns` for german with the reroute hook, `/de/ueber-uns` should return 200, but `/de/about` should be a 404. 

is it sufficient to have an `error(404)` in `getRoutePath(url,lang)` from my example above in case it does not find a localized name for that url? This would require that the function has mappings for all routes in all supported languages, which can be quite a few depending on the application.

**Comment by dominikg** [comment]
doing ssr with `<html lang="XX">` set is important regardless of i18n framework you end up using, as is knowing which lang is currently active during ssr and on the client. Are these generic and important enough that we provide first-class support for it in kit itself?

**Comment by Antonio-Bennett** [comment]
I thought the second part of the original proposal could alleviate this?

**Comment by dummdidumm** [comment]
> `/docs` and `/en/docs` are the same under the hood, `reroute` should allow to 404/redirect on one of them

You could return `/no-match` from `reroute` in case you detect that, and then no route will be found, leading to a 404. If you want to redirect instead, then that is the job of `handle`, where you have access to the original URL.

> Additionally, reroute currently only provides a one-way mapping of an incoming url to a route, but usually you need the opposite direction too to be able to produce matching hrefs during ssr.

That sounds like a bidirectional map which you can implement yourself as needed in userland. I don't think we should have something built-in here.

> People have also mentioned wanting to reroute using cookies, which the current API doesn't allow. E.g. you might put someone into an A/B test bucket and want to show them the same version of the page everytime.

What indeed would be benefitial is to have access to headers/cookies during server rendering. We recently added [server-side route resolution](https://svelte.dev/docs/kit/configuration#router) and plan on adding async reroute for `hooks.server.js` which would then retrieve headers/cookies.

It would be interesting to know where you plan on setting the cookie for the test bucket. Somewhere else in front of SvelteKit, i.e. some kind of middleware? Or would you want to do that within SvelteKit, which would mean you would need to set headers/cookies in `reroute` (because how else would you be able to handle the first visit case)?

**Comment by benmccann** [comment]
> plan on adding async reroute for hooks.server.js 

I wonder if people will want to implement reroute in a cookie-aware fashion on the server and non-cookie aware fashion on the client. And I wonder what happens if the provide an implementation for both the server hook and universal hook - does it run twice, what order do they run, etc.?

**Comment by benmccann** [comment]
> Or would you want to do that within SvelteKit, which would mean you would need to set headers/cookies in reroute (because how else would you be able to handle the first visit case)?

Yes, I would want to do that within SvelteKit. Something that's not clear to me from the docs currently is what URL handle sees (the original or rerouted). I think it might be nice if we added a property to the event for the other. I.e. if `url` is the original URL then also add a property for `reroutedURL` or if URL is the rerouted URL then a property for `originalURL`. The presence or absence of this property would allow you to solve that problem

**Comment by dummdidumm** [comment]
> Something that's not clear to me from the docs currently is what URL handle sees

`event.url` is the original one, and from `route.id` you can kind of infer where you're actually going, but you're right that currently there's no way to tell what the rerouted pathname actually is.

> I wonder if people will want to implement reroute in a cookie-aware fashion on the server and non-cookie aware fashion on the client. And I wonder what happens if the provide an implementation for both the server hook and universal hook - does it run twice, what order do they run, etc.?

The idea is: You can have either `reroute` in `hooks.js` or in `hooks.server.js`, but not both (it's a validation error). And if you want to use the one in `hooks.server.js` you have to use server-side route resolution, which means the backend is called for each route resolution (and rerun runs there).

> Returning a non-existing route from reroute to trigger a 404 feels like a workaround, error handling should be explicit in this case.

I start to think that `reroute` in itself is too inflexible for all the things you want to do with it. What you really want is to do these things in `handle` basically, where you have all the tools (redirects, throwing errors, doing something else entirely) at your disposal, and if reroute, or rather route resolution, would somehow be a part of the handle hook which you can invoke explicitly, then that could solve these advanced use cases.

**Comment by dummdidumm** [comment]
Since #13520 + #13549 you can use headers/cookies to make your decision where to go. If we want to have anything more than that my point in https://github.com/sveltejs/kit/pull/13422#issuecomment-3846149052 still stands - I think we would need a new hook that runs as a middleware, _before_ static pages even. I therefore propose to close this.

Update: we decided to close this for the reasons I mentioned.
